storm-core/src/jvm/org/apache/storm/stats/BoltExecutorStats.java

```diff
@@ -0,0 +1,107 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.storm.stats;
+
+import clojure.lang.PersistentVector;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.storm.metric.internal.MultiCountStatAndMetric;
+import org.apache.storm.metric.internal.MultiLatencyStatAndMetric;
+
+@SuppressWarnings("unchecked")
+public class BoltExecutorStats extends CommonStats {
+
+    public static final String ACKED = "acked";
+    public static final String FAILED = "failed";
+    public static final String EXECUTED = "executed";
+    public static final String PROCESS_LATENCIES = "process-latencies";
+    public static final String EXECUTE_LATENCIES = "execute-latencies";
+
+    public static final String[] BOLT_FIELDS = {ACKED, FAILED, EXECUTED, PROCESS_LATENCIES, EXECUTE_LATENCIES};
+
+    public BoltExecutorStats() {
+        super();
+
+        put(ACKED, new MultiCountStatAndMetric(NUM_STAT_BUCKETS));
+        put(FAILED, new MultiCountStatAndMetric(NUM_STAT_BUCKETS));
+        put(EXECUTED, new MultiCountStatAndMetric(NUM_STAT_BUCKETS));
+        put(PROCESS_LATENCIES, new MultiLatencyStatAndMetric(NUM_STAT_BUCKETS));
+        put(EXECUTE_LATENCIES, new MultiLatencyStatAndMetric(NUM_STAT_BUCKETS));
+    }
+
+    public MultiCountStatAndMetric getAcked() {
+        return (MultiCountStatAndMetric) this.get(ACKED);
+    }
+
+    public MultiCountStatAndMetric getFailed() {
+        return (MultiCountStatAndMetric) this.get(FAILED);
+    }
+
+    public MultiCountStatAndMetric getExecuted() {
+        return (MultiCountStatAndMetric) this.get(EXECUTED);
+    }
+
+    public MultiLatencyStatAndMetric getProcessLatencies() {
+        return (MultiLatencyStatAndMetric) this.get(PROCESS_LATENCIES);
+    }
+
+    public MultiLatencyStatAndMetric getExecuteLatencies() {
+        return (MultiLatencyStatAndMetric) this.get(EXECUTE_LATENCIES);
+    }
+
+    public void boltExecuteTuple(String component, String stream, long latencyMs) {
+        Object key = PersistentVector.create(component, stream);
+        this.getExecuted().incBy(key, this.rate);
+        this.getExecuteLatencies().record(key, latencyMs);
+    }
+
+    public void boltAckedTuple(String component, String stream, long latencyMs) {
+        Object key = PersistentVector.create(component, stream);
+        this.getAcked().incBy(key, this.rate);
+        this.getProcessLatencies().record(key, latencyMs);
+    }
+
+    public void boltFailedTuple(String component, String stream, long latencyMs) {
+        Object key = PersistentVector.create(component, stream);
+        this.getFailed().incBy(key, this.rate);
+
+    }
+
+    public Map renderStats() {
+        cleanupStats();
+        Map ret = new HashMap();
+        ret.putAll(valueStats(CommonStats.COMMON_FIELDS));
+        ret.putAll(valueStats(BoltExecutorStats.BOLT_FIELDS));
+        StatsUtil.putRawKV(ret, StatsUtil.TYPE, StatsUtil.KW_BOLT);
```

Comment:
`CommonStats.renderStats` (this function) was originally `stats/render-stats!` it is only called from executor.clj as the implementation of the `render-stats` method in an instance of `RunningExecutor` that it creates.

https://github.com/apache/storm/blob/cc1f6d77962ec96c69cf5af85e4016a9e08ccdcc/storm-core/src/clj/org/apache/storm/daemon/executor.clj#L408-L409

This in turn is only called by the `do-executor-heartbeat` function in worker.clj

https://github.com/apache/storm/blob/cc1f6d77962ec96c69cf5af85e4016a9e08ccdcc/storm-core/src/clj/org/apache/storm/daemon/worker.clj#L65-L81

The code there takes all of the executors from the worker calls `render-stats` on them, puts them into a map with the key being the executor id and adds that map into `zk-hb` under the `:executor-stats` keyword.

`zk-hb` is passed into `thriftify-zk-worker-hb` which pulls out `:executor-stats` filters out any executor that does not have stats and passes it to `thriftify-stats`

https://github.com/apache/storm/blob/cc1f6d77962ec96c69cf5af85e4016a9e08ccdcc/storm-core/src/clj/org/apache/storm/converter.clj#L255

which turns the map keys into `ExecutorInfo` instances and the map values into `ExecutorStats` instances. 

Why go to a clojure specific data structure that no one reads except to turn it into the java class we actually care about.

If we modify `renderStats` to return an `ExecutorStats` instance.  Then modify `thriftify-stats` to not modify the values in the map, just the keys, we should be good to go.  We will also need to update nimbus_test.clj because it too calls into `thriftify-zk-worker-hb` in a test.